tests/test_packages/test_skills/test_tac_participation/test_behaviours.py

```diff
@@ -0,0 +1,327 @@
+        self.game._phase = Phase.PRE_GAME
+
+        # operation
+        with patch.object(self.logger, "log"):
+            self.transaction_process_behaviour.act()
+
+        # after
+        assert self.game.phase == Phase.PRE_GAME
+        self.assert_quantity_in_outbox(0)
+
+    def test_act_ii(self):
+        """Test the act method of the transaction_process behaviour where phase is GAME."""
+        # setup
+        self.game._phase = Phase.GAME
+
+        self.skill.skill_context._agent_context._shared_state = {
+            "transactions": {
+                self.tx_ids[0]: {
+                    "terms": self.terms[0],
+                    "sender_signature": self.sender_signatures[0],
+                    "counterparty_signature": self.counterparty_signatures[0],
+                },
+                self.tx_ids[1]: {
+                    "terms": self.terms[1],
+                    "sender_signature": self.sender_signatures[1],
+                    "counterparty_signature": self.counterparty_signatures[1],
+                },
+            }
+        }
+
+        tac_dialogue = self.prepare_skill_dialogue(
+            self.tac_dialogues, self.list_of_tac_messages,
+        )
+        self.game._tac_dialogue = tac_dialogue
+
+        # operation
+        with patch.object(self.logger, "log") as mock_logger:
+            self.transaction_process_behaviour.act()
+
+        # after
+        self.assert_quantity_in_outbox(2)
+
+        # _process_transactions
+        count = 0
+        while count != len(self.terms):
+            mock_logger.assert_any_call(
+                logging.INFO, f"sending transaction {self.tx_ids[count]} to controller."
+            )
+            message = self.get_message_from_outbox()
+            has_attributes, error_str = self.message_has_attributes(
+                actual_message=message,
+                message_type=TacMessage,
+                performative=TacMessage.Performative.TRANSACTION,
+                to=COUNTERPARTY_ADDRESS,
+                sender=self.skill.skill_context.agent_address,
+                target=tac_dialogue.last_message.message_id
+                - 2
+                + count,  # 2 tx messages already added to the tac_dialogue
+                transaction_id=self.tx_ids[count],
+                ledger_id=self.terms[count].ledger_id,
+                sender_address=self.terms[count].sender_address,
+                counterparty_address=self.terms[count].counterparty_address,
+                amount_by_currency_id=self.terms[count].amount_by_currency_id,
+                fee_by_currency_id=self.terms[count].fee_by_currency_id,
+                quantities_by_good_id=self.terms[count].quantities_by_good_id,
+                sender_signature=self.sender_signatures[count],
+                counterparty_signature=self.counterparty_signatures[count],
+                nonce=self.terms[count].nonce,
+            )
+            assert has_attributes, error_str
+            count += 1
```

Comment:
what's that for? tests run in any order, so don't assume order across methods

Reply:
Yes. That is for the while loop. 

The skill code sends a message for every tx in shared_state. So here, the while loops through transactions and validates every message in the outbox.